src/client/components/builder/constants/MapTable.tsx

```diff
@@ -0,0 +1,189 @@
+      payload: {
+        currIndex: index,
+        element: { ...constant, table: constant.table },
+        configArrName: ConfigArrayEnum.Constants,
+      },
+    })
+  }
+
+  return (
+    <>
+      <HStack w="100%" alignItems="flex-start">
+        <Box fontSize="20px" pt={2}>
+          <BiTable />
+        </Box>
+        <VStack align="stretch" w="100%">
+          <Input
+            type="text"
+            placeholder="Table Name"
+            name="title"
+            onChange={handleTitleChange}
+            value={title}
+          />
+          <Table variant="simple">
+            <Thead>
+              <Tr>
+                <Th>Reference</Th>
+                <Th>Constant</Th>
+                <Th />
+              </Tr>
+            </Thead>
+          </Table>
+          {table.map(({ key, value }, index) => (
+            <HStack w="100%" alignItems="flex-start" key={index}>
+              <Input
+                type="text"
+                placeholder="Reference"
+                name="Reference"
+                onChange={(e) => {
+                  handleUpdateTableRow({ key: e.target.value, value }, index)
+                }}
+                value={key}
+                w="45%"
+              />
+              <Input
+                type="number"
+                placeholder="Constant"
+                name="constant"
+                onChange={(e) => {
+                  handleUpdateTableRow(
+                    { key, value: Number(e.target.value) },
+                    index
+                  )
+                }}
+                value={value}
+                w="45%"
+              />
+              <IconButton
+                borderRadius={0}
+                variant="ghost"
+                aria-label="delete item"
+                fontSize="20px"
+                icon={<BiTrash />}
+                onClick={() => handleDeleteTableRow(index)}
+              />
+            </HStack>
+          ))}
+          <Button
+            leftIcon={<BiPlusCircle />}
+            variant="outline"
+            colorScheme="primary"
+            aria-label="add map item"
+            onClick={handleAddTableRow}
+            w="91.25%"
+          >
+            Add map constant
+          </Button>
+        </VStack>
+      </HStack>
+    </>
+  )
+}
+
+const PreviewComponent: ConstantFieldComponent = ({ constant }) => {
+  const { title, table } = constant
+  return (
+    <VStack align="stretch" w="100%" spacing={6}>
+      <VStack align="stretch">
+        <HStack>
+          <BiTable fontSize="20px" />
+          <Text>{title}</Text>
+        </HStack>
+        <HStack>
+          <Table variant="simple">
+            <Thead>
+              <Tr>
+                <Th>Reference</Th>
+                <Th>Constant</Th>
+                <Th />
+              </Tr>
+            </Thead>
+            <Tbody>
+              {table.map(({ key, value }, index) => (
+                <Tr key={index}>
```

Comment:
```suggestion
                key && <Tr key={index}>
```

Should we skip rendering the entire row if key is also empty?

Reply:
I think we should still render it, because as of now, users can fill in only the `value` but not the `key`. And they might not realize what the issue is if we throw an error but hide the missing row:
<img width="891" alt="Screenshot 2021-02-26 at 5 28 39 PM" src="https://user-images.githubusercontent.com/19917616/109282215-34030380-7858-11eb-9f78-2aaa0c91fd51.png">
<img width="903" alt="Screenshot 2021-02-26 at 5 28 43 PM" src="https://user-images.githubusercontent.com/19917616/109282225-36fdf400-7858-11eb-8f2b-e799609ccc74.png">